--- conf/chilli.in	2015-08-28 10:38:11.665845630 -0400
+++ conf/chilli.in	2015-08-28 10:36:13.531241373 -0400
@@ -43,82 +45,148 @@
     exit 0
 }
 
+readconfig() {
+        chilliconf="/etc/chilli/config"
+        if [ -f "/etc/chilli/config.new" ]; then
+            mv /etc/chilli/config.new /etc/chilli/config
+        fi
 
-start() {
-    echo -n $"Starting $prog: "
-
-    check_required
+        DNS1=`grep nameserver /etc/resolv.conf | head -1 | cut -d" " -f2`
+        DNS2=`grep nameserver /etc/resolv.conf | tail -1 | cut -d" " -f2`
+        sed -i "s/^HS_DNS1.*/HS_DNS1=$DNS1/g" $chilliconf
+        sed -i "s/^HS_DNS2.*/HS_DNS2=$DNS2/g" $chilliconf
+
+        GW=`grep GATEWAY /etc/sysconfig/network-scripts/ifcfg-eth1 | head -1 | cut -d"=" -f2`
+        WAN="eth0"
+        LAN="eth1"
+        if [ -n $GW ]; then GW=`echo "$GW" | tr -d '"'` ; fi
+        if [ ! -z $GW ]; then
+                WAN="eth1"
+                LAN="eth0"
+                IP=`grep IPADDR /etc/sysconfig/network-scripts/ifcfg-eth0 | head -1 | cut -d"=" -f2 | tr -d '"'`
+                NETMASK=`grep NETMASK /etc/sysconfig/network-scripts/ifcfg-eth0 | head -1 | cut -d"=" -f2 | tr -d '"'`
+        else
+                GW=`grep GATEWAY /etc/sysconfig/network-scripts/ifcfg-eth0 | head -1 | cut -d"=" -f2`
+                IP=`grep IPADDR /etc/sysconfig/network-scripts/ifcfg-eth1 | head -1 | cut -d"=" -f2 | tr -d '"'`
+                NETMASK=`grep NETMASK /etc/sysconfig/network-scripts/ifcfg-eth1 | head -1 | cut -d"=" -f2 | tr -d '"'`
+        fi
+        sed -i "s/^HS_WANIF=.*/HS_WANIF=$WAN/g" $chilliconf
+        sed -i "s/^HS_LANIF=.*/HS_LANIF=$LAN/g" $chilliconf
 
-    /sbin/modprobe tun >/dev/null 2>&1
-    echo 1 > /proc/sys/net/ipv4/ip_forward
+        iparr=(${IP//./ })
+        maskarr=(${NETMASK//./ })
+        netarr=()
+        for i in {0..3}
+        do
+            netarr[$i]=$((${iparr[i]}&${maskarr[i]}))
+        done
+        NET=$(printf ".%s" "${netarr[@]}")
+        NET=${NET:1}
+
+        sed -i "s/^HS_NETWORK.*/HS_NETWORK=$NET/g" $chilliconf
+        sed -i "s/^HS_NETMASK.*/HS_NETMASK=$NETMASK/g" $chilliconf
+        sed -i "s/^HS_UAMLISTEN.*/HS_UAMLISTEN=$IP/g" $chilliconf
+
+        ipsplit=`echo ${IP//./ }`
+        iphex=`printf "%02X" $ipsplit `
+        tlv="2B063304"$iphex
+        sed -i "s/^HS_DHCPOPT.*/HS_DHCPOPT=\"$tlv\"/g" $chilliconf
 
-    [ -e /dev/net/tun ] || {
-        (cd /dev; mkdir net; cd net; mknod tun c 10 200)
-    }
-
-    writeconfig
-    radiusconfig
-
-    test ${HS_ADMINTERVAL:-0} -gt 0 && {
-        (crontab -l 2>&- | grep -v $0
-        echo "*/$HS_ADMINTERVAL * * * * $0 radconfig") | crontab - 2>&-
-    }
-
-    test ${HS_LANIF_KEEPADDR:-0} -eq 0 && {
-        storeroutes $ROUTES_FILE
-        ifconfig $HS_LANIF 0.0.0.0
-    }
-
-    daemon $exec -c $CONFIG
-    retval=$?
-    echo
-    [ $retval -eq 0 ] && touch $lockfile
-    return $retval
+        . $chilliconf
+        ROUTES_FILE=/var/run/chilli_routes.$HS_LANIF
 }
 
-stop() {
-    echo -n $"Stopping $prog: "
-    crontab -l 2>&- | grep -v $0 | crontab -
-
-    test ${HS_LANIF_KEEPADDR:-0} -eq 0 && {
-        ifconfig $HS_LANIF $HS_UAMLISTEN netmask $HS_NETMASK
-    }
-
-    killproc $prog
-    retval=$?
-    echo
-    [ $retval -eq 0 ] && rm -f $lockfile
-    [ $retval -eq 0 ] && rm -f $PIDFILE
-    [ $retval -eq 0 ] && rm -f $RUN_D/$IPCFILE
-    [ $retval -eq 0 ] && rm -f $CMDSOCK
-    [ $retval -eq 0 ] && /bin/rm -f @VARRUN@/chilli.*.cfg.bin
-
-    test ${HS_LANIF_KEEPADDR:-0} -eq 0 && {
-        restoreroutes $ROUTES_FILE
-        [ -f $ROUTES_FILE ] && rm $ROUTES_FILE
-    }
+case "$1" in
+    start)
+        if [ '0' == `uci -q get chilli.setting.enable` ]; then
+            exit 0;
+        fi
 
-    return $retval
-}
+        echo -n $"Starting $prog: "
 
-restart() {
-    stop
-    start
-}
+        check_required
 
-case "$1" in
-    start|stop|restart)
-        $1
+        /sbin/modprobe tun >/dev/null 2>&1
+        echo 1 > /proc/sys/net/ipv4/ip_forward
+
+        [ -e /dev/net/tun ] || {
+            (cd /dev; mkdir net; cd net; mknod tun c 10 200)
+        }
+
+        readconfig
+        writeconfig
+        radiusconfig
+
+        test ${HS_ADMINTERVAL:-0} -gt 0 && {
+            (crontab -l 2>&- | grep -v $0
+            echo "*/$HS_ADMINTERVAL * * * * $0 radconfig") | crontab - 2>&-
+        }
+
+        test ${HS_LANIF_KEEPADDR:-0} -eq 0 && {
+            storeroutes $ROUTES_FILE
+            ifconfig $HS_LANIF 0.0.0.0
+        }
+
+        if [ "$(which start-stop-daemon 2>/dev/null)" = "" ]; then
+            daemon $exec -c $CONFIG --pidfile=$pidfile
+        else
+            start-stop-daemon -S --pidfile=$pidfile --user=chilli \
+                    --exec $exec -- -c $CONFIG &
+        fi
+        retval=$?
+        echo
+        [ $retval -eq 0 ] && touch $lockfile
+
+        if [ -z $2 ]; then
+            /etc/init.d/xl2tpd restart
+        fi
+        ;;
+    stop)
+        echo -n $"Stopping $prog: "
+        crontab -l 2>&- | grep -v $0 | crontab -
+
+        if [ "$(which start-stop-daemon 2>/dev/null)" != "" ]; then
+            start-stop-daemon -K --pidfile=$pidfile --user=chilli \
+                $exec
+        fi
+
+        killproc $prog
+        retval=$?
+        echo
+        [ $retval -eq 0 ] && rm -f $lockfile
+        [ $retval -eq 0 ] && rm -f $PIDFILE
+        [ $retval -eq 0 ] && rm -f $RUN_D/$IPCFILE
+        [ $retval -eq 0 ] && rm -f $CMDSOCK
+        [ $retval -eq 0 ] && /bin/rm -f @VARRUN@/chilli.*.cfg.bin
+
+        test ${HS_LANIF_KEEPADDR:-0} -eq 0 && {
+            ifconfig $HS_LANIF $HS_UAMLISTEN netmask $HS_NETMASK
+            restoreroutes $ROUTES_FILE
+            [ -f $ROUTES_FILE ] && rm $ROUTES_FILE
+        }
+
+        if [ -z $2 ]; then
+            /etc/init.d/xl2tpd restart
+        fi
+        ;;
+    restart)
+        $0 stop me
+        sleep 1
+        $0 start me
+        RETVAL=$?
+        if [ -z $2 ]; then
+            /etc/init.d/xl2tpd restart
+        fi
         ;;
     force-reload)
-        restart
+        $0 restart
         ;;
     status)
         status $prog
         ;;
     try-restart|condrestart)
         if status $prog >/dev/null ; then
-            restart
+            $0 restart
         fi
 	;;
     reload)
@@ -130,6 +198,7 @@
         action $"Service ${0##*/} does not support the reload action: " /bin/false
         exit 3
         ;;
+
     *)
         echo $"Usage: $0 {start|stop|status|restart|try-restart|force-reload}"
         exit 2
